src/python/pants/backend/python/lint/mypy/rules.py

```diff
@@ -61,6 +70,41 @@ def generate_args(mypy: MyPy, *, file_list_path: str) -> Tuple[str, ...]:
     return tuple(args)
 
 
+# MyPy searches for types for a package in packages containing a `py.types` marker file or else in
+# a sibling `<package>-stubs` package as per PEP-0561. Going further than that PEP, MyPy restricts
+# its search to `site-packages`. Since PEX deliberately isolates itself from `site-packages` as
+# part of its raison d'être, we monkey-patch `site.getsitepackages` to look inside the scrubbed
+# PEX sys.path before handing off to `mypy`.
+#
+# As a complication, MyPy does its own validation to ensure packages aren't both available in
+# site-packages and on the PYTHONPATH. As such, we elide all PYTHONPATH entries from artificial
+# site-packages we set up since MyPy will manually scan PYTHONPATH outside this PEX to find
+# packages.
+#
+# See:
+#   https://mypy.readthedocs.io/en/stable/installed_packages.html#installed-packages
+#   https://www.python.org/dev/peps/pep-0561/#stub-only-packages
+LAUNCHER_FILE = FileContent(
+    "__pants_mypy_launcher.py",
+    dedent(
+        """\
+        import os
+        import runpy
+        import site
+        import sys
+        PYTHONPATH = frozenset(
+            os.path.realpath(p)
+            for p in os.environ.get('PYTHONPATH', '').split(os.pathsep)
+        )
+        site.getsitepackages = lambda: [
+            p for p in sys.path if os.path.realpath(p) not in PYTHONPATH
+        ]
+        runpy.run_module('mypy', run_name='__main__')
```

Comment:
For robustness, you'll need to monkeypatch `site.getusersitepackages = lambda: ''` as well because:
https://github.com/python/mypy/blob/e8e44e6f9f4a865802d544ea48ea765dfeccdd67/mypy/sitepkgs.py#L24

Right now that check fails since our very old version of virtualenv creates neither `site.getsitepackages` nor `site.getusersitepackages`. I'll prepare a fix for that if you want to take this one home.

Reply:
It would probably be better to map getusersitepackages to a known empty dir, but `''` (CWD) probably suffices.